From 2d59f5d6b9b574029a02b9de5d4c2ae759272bea Mon Sep 17 00:00:00 2001
From: Peter Delevoryas <peter@pjd.dev>
Date: Wed, 24 Aug 2022 01:55:12 -0700
Subject: [PATCH 512/532] hw/i2c/aspeed: Fix bus derivation for slave events

The existing code assumes there's one AspeedI2CBus attached to each
I2CBus. That's not the case in a multi-SoC system where two
AspeedI2CBus's share a single I2CBus to communicate with each other.

Signed-off-by: Peter Delevoryas <peter@pjd.dev>
---
 hw/i2c/aspeed_i2c.c         | 7 +++----
 include/hw/i2c/aspeed_i2c.h | 1 +
 2 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/hw/i2c/aspeed_i2c.c b/hw/i2c/aspeed_i2c.c
index 33c3b27375..089fd9e75d 100644
--- a/hw/i2c/aspeed_i2c.c
+++ b/hw/i2c/aspeed_i2c.c
@@ -1147,8 +1147,7 @@ static int aspeed_i2c_bus_new_slave_event(AspeedI2CBus *bus,
 
 static int aspeed_i2c_bus_slave_event(I2CSlave *slave, enum i2c_event event)
 {
-    BusState *qbus = qdev_get_parent_bus(DEVICE(slave));
-    AspeedI2CBus *bus = ASPEED_I2C_BUS(qbus->parent);
+    AspeedI2CBus *bus = ASPEED_I2C_BUS_SLAVE(slave)->bus;
     uint32_t reg_intr_sts = aspeed_i2c_bus_intr_sts_offset(bus);
     uint32_t reg_byte_buf = aspeed_i2c_bus_byte_buf_offset(bus);
     uint32_t reg_dev_addr = aspeed_i2c_bus_dev_addr_offset(bus);
@@ -1206,8 +1205,7 @@ static void aspeed_i2c_bus_new_slave_send_async(AspeedI2CBus *bus, uint8_t data)
 
 static void aspeed_i2c_bus_slave_send_async(I2CSlave *slave, uint8_t data)
 {
-    BusState *qbus = qdev_get_parent_bus(DEVICE(slave));
-    AspeedI2CBus *bus = ASPEED_I2C_BUS(qbus->parent);
+    AspeedI2CBus *bus = ASPEED_I2C_BUS_SLAVE(slave)->bus;
     uint32_t reg_intr_sts = aspeed_i2c_bus_intr_sts_offset(bus);
     uint32_t reg_byte_buf = aspeed_i2c_bus_byte_buf_offset(bus);
 
@@ -1265,6 +1263,7 @@ static void aspeed_i2c_bus_realize(DeviceState *dev, Error **errp)
     s->bus = i2c_init_bus(dev, name);
     s->slave = i2c_slave_create_simple(s->bus, TYPE_ASPEED_I2C_BUS_SLAVE,
                                        0xff);
+    ASPEED_I2C_BUS_SLAVE(s->slave)->bus = s;
 
     memory_region_init_io(&s->mr, OBJECT(s), &aspeed_i2c_bus_ops,
                           s, name, aic->reg_size);
diff --git a/include/hw/i2c/aspeed_i2c.h b/include/hw/i2c/aspeed_i2c.h
index eacf1dd0eb..b9165fbe9d 100644
--- a/include/hw/i2c/aspeed_i2c.h
+++ b/include/hw/i2c/aspeed_i2c.h
@@ -269,6 +269,7 @@ struct AspeedI2CState {
 OBJECT_DECLARE_SIMPLE_TYPE(AspeedI2CBusSlave, ASPEED_I2C_BUS_SLAVE)
 struct AspeedI2CBusSlave {
     I2CSlave i2c;
+    AspeedI2CBus *bus;
 };
 
 struct AspeedI2CClass {
-- 
2.44.2

